src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```
@@ -0,0 +1,377 @@
+            {
+                rankOrModifier = ByRef;
+                return true;
+            }
+
+            if (TryStripFirstCharAndTrailingSpaces(ref input, '['))
+            {
+                // SZArray := []
+                // MDArray := [*] or [,] or [,,,, ...]
+
+                int rank = 1;
+                bool hasSeenAsterisk = false;
+
+            ReadNextArrayToken:
+
+                if (TryStripFirstCharAndTrailingSpaces(ref input, ']'))
+                {
+                    // End of array marker
+                    rankOrModifier = rank == 1 && !hasSeenAsterisk ? SZArray : rank;
+                    return true;
+                }
+
+                if (!hasSeenAsterisk)
+                {
+                    if (rank == 1 && TryStripFirstCharAndTrailingSpaces(ref input, '*'))
+                    {
+                        // [*]
+                        hasSeenAsterisk = true;
+                        goto ReadNextArrayToken;
+                    }
+                    else if (TryStripFirstCharAndTrailingSpaces(ref input, ','))
+                    {
+                        // [,,, ...]
+                        checked { rank++; }
+                        goto ReadNextArrayToken;
+                    }
+                }
+
+                // Don't know what this token is.
+                // Fall through to 'return false' statement.
+            }
+
+            input = originalInput; // ensure 'ref input' not mutated
+            rankOrModifier = 0;
+            return false;
+        }
+
+        internal static bool TryStripFirstCharAndTrailingSpaces(ref ReadOnlySpan<char> span, char value)
+        {
+            if (!span.IsEmpty && span[0] == value)
+            {
+                span = span.Slice(1).TrimStart();
+                return true;
+            }
+            return false;
+        }
+
+        internal static InvalidOperationException InvalidOperation_MaxNodesExceeded(int limit) =>
```

It would be better for these methods to actually throw instead of returning the exception. It is the preferred ThrowHelper pattern going forward.